### ISSUE TITLE
[BUG] SKImage.FromEncodedData fails on loading image

### ISSUE THREAD

**Issue by Spiralis** [mono/SkiaSharp#1249]
**Description**

I am getting an exception when trying to load a botmap from file into an AvaloniaUI `<Image>` control.  The bitmap is a 1 pixel wide image, but with a potentially large (ridiculous) height. The image control will scale the image to fit the associated size. 

It works when the image is not too high. But when it gets large it fails. 

**Code**

1. The image is created using SixLabors ImageSharp, and was saved to disk via `img.Save("name.bmp", new BmpEncoder());`. The file appears on disk and can be opened in image-viewers without any issue. 

The size for the **failing** test-scenario is about 9 MB (1 x 2 309 534 pixels). 

The size for a **non-failing** test-scenario is about 4 KB (1 x 1 065 pixels).

2. An Avalonia UI `<Image>` control is set up as follows
```
<Image Width="10" Stretch="Fill" DockPanel.Dock="Right" />
```
When the view-model detects that the file is created it assigns the filename to the image control's `Source` attribute.

**Expected Behavior**

An image that expands in width to 10 pixels (based on the single pixel width), with a height equal to the height of the DockPanel. The height should (normally) shrink to fit the height.

For the working test-scenario this works exactly as expected.

**Actual Behavior**

I get an exception as a result of assigning the filename to the Source attribute for the failing test-scenario.

```
System.ArgumentException
  HResult=0x80070057
  Message=Unable to load bitmap from provided data
  Source=Avalonia.Skia
  StackTrace:
   at Avalonia.Skia.ImmutableBitmap..ctor(Stream stream)
   at Avalonia.Skia.PlatformRenderInterface.LoadBitmap(Stream stream)
   at Avalonia.Skia.PlatformRenderInterface.LoadBitmap(String fileName)
   at Avalonia.Media.Imaging.Bitmap..ctor(String fileName)
   at LogExpress.ViewModels.LogViewModel.<.ctor>b__15_13(String fileName) in D:\src\Haive\LogExpress\src\LogExpress-Avalonia\ViewModels\LogViewModel.cs:line 75
   at System.Reactive.AnonymousSafeObserver`1.OnNext(T value) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\AnonymousSafeObserver.cs:line 54
   at System.Reactive.Linq.ObservableImpl.DistinctUntilChanged`2._.OnNext(TSource value) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\DistinctUntilChanged.cs:line 74
   at System.Reactive.Linq.ObservableImpl.Switch`1._.InnerObserver.OnNext(TSource value) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Switch.cs:line 100
   at System.Reactive.IdentitySink`1.OnNext(T value) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\IdentitySink.cs:line 16
   at System.Reactive.Linq.ObservableImpl.Cast`2._.OnNext(TSource value) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Cast.cs:line 41
   at System.Reactive.IdentitySink`1.OnNext(T value) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\IdentitySink.cs:line 16
   at System.Reactive.Subjects.Subject`1.OnNext(T value) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Subjects\Subject.cs:line 148
   at System.Reactive.IdentitySink`1.OnNext(T value) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\IdentitySink.cs:line 16
   at System.Reactive.Linq.ObservableImpl.SelectMany`2.EnumerableSelector._.OnNext(TSource value) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\SelectMany.cs:line 1351
   at System.Reactive.Linq.ObservableImpl.Buffer`2.Boundaries._.BufferClosingObserver.OnNext(TBufferClosing value) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Buffer.cs:line 829
   at System.Reactive.Linq.ObservableImpl.Merge`1.Observables._.InnerObserver.OnNext(TSource value) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Merge.cs:line 236
   at System.Reactive.Linq.ObservableImpl.Select`2.Selector._.OnNext(TSource value) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Select.cs:line 48
   at System.Reactive.Subjects.Subject`1.OnNext(T value) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Subjects\Subject.cs:line 148
   at ReactiveUI.IReactiveObjectExtensions.ExtensionState`1.NotifyObservable[T](TSender rxObj, T item, ISubject`1 subject) in d:\a\1\s\src\ReactiveUI\ReactiveObject\IReactiveObjectExtensions.cs:line 386
   at ReactiveUI.IReactiveObjectExtensions.ExtensionState`1.RaisePropertyChanged(String propertyName) in d:\a\1\s\src\ReactiveUI\ReactiveObject\IReactiveObjectExtensions.cs:line 380
   at ReactiveUI.IReactiveObjectExtensions.RaisingPropertyChanged[TSender](TSender reactiveObject, String propertyName) in d:\a\1\s\src\ReactiveUI\ReactiveObject\IReactiveObjectExtensions.cs:line 193
   at ReactiveUI.IReactiveObjectExtensions.RaiseAndSetIfChanged[TObj,TRet](TObj reactiveObject, TRet& backingField, TRet newValue, String propertyName) in d:\a\1\s\src\ReactiveUI\ReactiveObject\IReactiveObjectExtensions.cs:line 119
   at LogExpress.Services.VirtualLogFile.set_LogLevelMapFile(String value) in D:\src\Haive\LogExpress\src\LogExpress-Avalonia\Services\VirtualLogFile.cs:line 194
   at LogExpress.Services.VirtualLogFile.CreateLogLevelMapImage() in D:\src\Haive\LogExpress\src\LogExpress-Avalonia\Services\VirtualLogFile.cs:line 185
   at LogExpress.Services.VirtualLogFile.<.ctor>b__11_5(Boolean x) in D:\src\Haive\LogExpress\src\LogExpress-Avalonia\Services\VirtualLogFile.cs:line 134
   at System.Reactive.AnonymousSafeObserver`1.OnNext(T value) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\AnonymousSafeObserver.cs:line 44
   at System.Reactive.ObserveOnObserverNew`1.DrainStep(ConcurrentQueue`1 q) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\ScheduledObserver.cs:line 577
   at System.Reactive.ObserveOnObserverNew`1.DrainShortRunning(IScheduler recursiveScheduler) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\ScheduledObserver.cs:line 509
   at System.Reactive.ObserveOnObserverNew`1.<>c.<.cctor>b__17_0(IScheduler scheduler, ObserveOnObserverNew`1 self) in d:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\ScheduledObserver.cs:line 497
   at Avalonia.Threading.AvaloniaScheduler.<>c__DisplayClass2_1`1.<Schedule>b__1()
   at Avalonia.Threading.JobRunner.RunJobs(Nullable`1 priority)
   at Avalonia.Win32.Win32Platform.WndProc(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
   at Avalonia.Win32.Interop.UnmanagedMethods.DispatchMessage(MSG& lpmsg)
   at Avalonia.Win32.Win32Platform.RunLoop(CancellationToken cancellationToken)
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken)
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args)
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime[T](T builder, String[] args, ShutdownMode shutdownMode)
   at LogExpress.Program.Main(String[] args) in D:\src\Haive\LogExpress\src\LogExpress-Avalonia\Program.cs:line 32

  This exception was originally thrown at this call stack:
    Avalonia.Skia.ImmutableBitmap.ImmutableBitmap(System.IO.Stream)
    Avalonia.Skia.PlatformRenderInterface.LoadBitmap(System.IO.Stream)
    Avalonia.Skia.PlatformRenderInterface.LoadBitmap(string)
    Avalonia.Media.Imaging.Bitmap.Bitmap(string)
    LogExpress.ViewModels.LogViewModel..ctor.AnonymousMethod__15_13(string) in LogViewModel.cs
    System.Reactive.AnonymousSafeObserver<T>.OnNext(T) in AnonymousSafeObserver.cs
    System.Reactive.Linq.ObservableImpl.DistinctUntilChanged<TSource, TKey>._.OnNext(TSource) in DistinctUntilChanged.cs
    System.Reactive.Linq.ObservableImpl.Switch<TSource>._.InnerObserver.OnNext(TSource) in Switch.cs
    System.Reactive.IdentitySink<T>.OnNext(T) in IdentitySink.cs
    System.Reactive.Linq.ObservableImpl.Cast<TSource, TResult>._.OnNext(TSource) in Cast.cs
    ...
    [Call Stack Truncated]
```

**Basic Information**

- Version with issue:  AvaloniaUI v9.9 (not sure which version of SkiaSharp is used for it)
- Last known good version:  Don't know. This is new code.
- IDE:  Visual Studio
- Platform Target Frameworks: 
  - Windows Classic:  Windows 10

**Screenshots**

[failing-image.zip](https://github.com/mono/SkiaSharp/files/4533900/failing-image.zip)
[non-failing-image.zip](https://github.com/mono/SkiaSharp/files/4533904/non-failing-image.zip)

**Reproduction Link**

N/A

**Comment by Spiralis** [comment]
@kekekeks (from the AvaloniaUI core team I believe) suggested that this could be an issue in SkiaSharp and that I should file an issue for it. 

He didn't think that they could fix this by changing anything within Avalonia.

**Comment by kekekeks** [comment]
Relevant image loading code:

```cs
        public ImmutableBitmap(Stream stream)
        {
            using (var skiaStream = new SKManagedStream(stream))
            {
                using (var data = SKData.Create(skiaStream))
                    _image = SKImage.FromEncodedData(data);

                if (_image == null)
                {
                    throw new ArgumentException("Unable to load bitmap from provided data");
                }

                PixelSize = new PixelSize(_image.Width, _image.Height);

                // TODO: Skia doesn't have an API for DPI.
                Dpi = new Vector(96, 96);
            }
        }
```
https://github.com/AvaloniaUI/Avalonia/blob/9b73c1027b8bc03b8f2501d0fe4df8d311be7dac/src/Skia/Avalonia.Skia/ImmutableBitmap.cs#L20-L37

**Comment by Gillibald** [comment]
Skia has a limit for bitmap sizes. Can't remember the actual value.

**Comment by Gillibald** [comment]
https://groups.google.com/forum/m/#!topic/skia-discuss/886WuVcQUGo

**Comment by Spiralis** [comment]
Why does it have this limitation, by the way? 

I can make them multiple images and create new Image controls to handle each, but I would have to handle scaling myself. Which is a pity when the single image control handles this nicely. I am not even sure how to stack them and scale them, so that will mean a lot of additional work.

Does this btw mean that any Avalonia <Image> control will not be able to handle images larger than 32Kx32K?

Finally, why the 32Kx32K limitations? Would it not make sense if the limitation was by the multiplied total, allowing for rectangles to go beyond 32K in one direction? Or is this related to some hardware specific graphics card limitations?

**Comment by mattleibow** [comment]
@Spiralis check out the link from @Gillibald for more info. It also has some info on drawing tiles, which may be something that a control suite can handle for you.

The issue might be loading this image to process... I'll see if I can whip up some demo code that makes it possible to load infinity without breaking... `SKCodec` hopefully supports reading chunks of pixels even if the full image can't be decoded in one go.

**Comment by mattleibow** [comment]
Duplicate of 648

**Comment by mattleibow** [comment]
Just so that we know, even macOS can't open it. The other one is fine, just the super huge one is broken. Windows is good.

<img width="431" alt="Screenshot 2020-04-26 at 13 36 23" src="https://user-images.githubusercontent.com/1096616/80306325-ee525600-87c2-11ea-8044-b5aa91674bd6.png">

**Comment by Gillibald** [comment]
It is just wrong having a bitmap that is that big. Think about some tiling mechanism and load your pixel data on-demand in chunks.

**Comment by mattleibow** [comment]
Yeah. This is going to explode your computer 2M x 2M pixels is like 4 trillion pixels - each with 32 bits of data. A HUGE amount of data to load into memory. 128 trillion bits, 16 trillion bytes, 2 TB or something. That is not going to load.

**Comment by mattleibow** [comment]
Just saw this about .bmp files: https://github.com/google/skia/blob/master/src/codec/SkBmpCodec.cpp#L286-L290

> ```cpp
> // Arbitrary maximum. Matches Chromium.
> constexpr int kMaxDim = 1 << 16;
> if (width <= 0 || height <= 0 || width >= kMaxDim || height >= kMaxDim) {
>     SkCodecPrintf("Error: invalid bitmap dimensions.\n");
>     return kInvalidInput;
> }
>```

This means I can't even instantiate the codec to load tiles of the bitmap... Let me try png... 🤞

**Comment by Spiralis** [comment]
Yeah. I have changed it to do chunks at a time. I used it to create a "heatmap" that I then scaled down. Problem was that each line needed to be represented. So, now I read a given number of lines, create the image, scale it down, and then processes the next chunk, and scales that etc. Then, at then end each of these images are tiled, but each of them are also scaled down. Because, the number of them is dynamic and have to handle any number of files/lines that I analyze. Which means that, if the number gets really large, I'd even need to do another pass on the generated images, combining a number of them and scaling again...

But, this surely can be closed,I think. I have solved it for my case at least.
